src/bpf/bpfctl.c

```diff
@@ -0,0 +1,298 @@
+}
+
+static int query_and_print_bpf_maps(void) {
+        _cleanup_(varlink_unrefp) Varlink *vl = NULL;
+        _cleanup_(table_unrefp) Table *map_table = NULL;
+        JsonVariant *i = NULL, *o = NULL;
+        int r;
+
+        map_table = table_new("name", "type", "memlock");
+        if (!map_table)
+                return log_oom();
+
+        r = varlink_connect_address(&vl, "/run/systemd/bpf/io.systemd.Bpf");
+        if (r < 0)
+                return log_error_errno(r, "Unable to connect to /run/systemd/bpf/io.systemd.Bpf: %m");
+
+        r = varlink_collect_full(vl, "io.systemd.Bpf.DescribeMaps", i, &o, NULL, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to collect varlink method: %m");
+
+        JSON_VARIANT_ARRAY_FOREACH(i, o) {
+                static const JsonDispatch dispatch_table[] = {
+                        { "name",    JSON_VARIANT_STRING,   json_dispatch_string, offsetof(struct BpfMapData, name),    JSON_MANDATORY },
+                        { "type",    JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct BpfMapData, type),    JSON_MANDATORY },
+                        { "memlock", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct BpfMapData, memlock), JSON_MANDATORY },
+                        {}
+                };
+                _cleanup_(bpf_map_data_done) BpfMapData p = {};
+                // _cleanup_free_ const char *map_type = NULL;
+
+                r = json_dispatch(i, dispatch_table, 0, &p);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to dispatch map table: %m");
+
+                // TODO: coredumps
+                // map_type = sym_libbpf_bpf_map_type_str(p.type);
+                // if (!map_type)
+                //         return log_error_errno(r, "Failed to translate bpf_map_type to a valid string: %m");
+
+                r = table_add_many(map_table,
+                                TABLE_STRING, p.name,
+                        //    TABLE_STRING, map_type,
+                                TABLE_UINT64, p.type,
+                                TABLE_SIZE, p.memlock);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to add to map table: %m");
+        }
+
+        r = table_print_with_pager(map_table, arg_json_format_flags, arg_pager_flags, /*show_header*/true);
+        if (r < 0)
+                return log_error_errno(r, "Failed to output map table: %m");
+
+        return 0;
+}
+
+static int query_and_print_bpf_progs(void) {
+        _cleanup_(varlink_unrefp) Varlink *vl = NULL;
+        _cleanup_(table_unrefp) Table *prog_table = NULL;
+        JsonVariant *i = NULL, *o = NULL;
+        int r;
+
+        prog_table = table_new("name", "type", "memlock", "tag", "map names", "run time ns", "run cnt");
+        if (!prog_table)
+                return log_oom();
+
+        r = varlink_connect_address(&vl, "/run/systemd/bpf/io.systemd.Bpf");
```

Comment:
This string is already defined in the daemon. I would prefer to have a `#define BPFD_VALINK_ADDRESS "…"` in a header somewhere.

Reply:
Sure, I defined it in bpfd-manager.h.